text/0000-supertrait-item-shadowing-v2.md

```diff
@@ -0,0 +1,152 @@
+- Start Date: 2024-05-04
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+When name resolution encounters an ambiguity between 2 trait methods when both traits are in scope, if one trait is a sub-trait of the other then select that method instead of reporting an ambiguity error.
+
+# Motivation
+[motivation]: #motivation
+
+
+The libs-api team would like to stabilize `Iterator::intersperse` but has a problem. The `itertools` crate already has:
+
+```rust
+// itertools
+trait Itertools: Iterator {
+    fn intersperse(self, element: Self::Item) -> Intersperse<Self>;
+}
+```
+
+This method is used in crates with code similar to the following:
+
+```rust
+use core::iter::Iterator; // Implicit import from prelude
+
+use itertools::Itertools as _;
+
+fn foo() -> impl Iterator<Item = &'static str> {
+    "1,2,3".split(",").intersperse("|")
+    // ^ This is ambiguous: it could refer to Iterator::intersperse or Itertools::intersperse
+}
+```
+
+This code actually works today because `intersperse` is an unstable API, which works because the compiler already has [logic](https://github.com/rust-lang/rust/pull/48552) to prefer stable methods over unstable methods when an amiguity occurs.
+
+Attempts to stabilize `intersperse` have failed with a large number of regressions [reported by crater](https://github.com/rust-lang/rust/issues/88967) which affect many popular crates. Even if these were to be manually corrected (since ambiguity is considered allowed breakage) we would have to go through this whole process again every time a method from `itertools` is uplifted to the standard library.
+
+# Proposed solution
+[proposed-solution]: #proposed-solution
+
+This RFC proposes to change name resolution to resolve the ambiguity in the following specific circumstances:
+- All method candidates are trait methods. (Inherent methods are already prioritized over trait methods)
+- One trait is transitively a sub-trait of all other traits in the candidate list.
+
+When this happens, the sub-trait method is selected instead of reporting an ambiguity error.
+
+Note that this only happens when *both* traits are in scope since this is required for the ambiguity to occur in the first place.
+
+When an ambiguity is resolved in this way, a lint warning is also emitted to warn the user about the potential ambiguity. The aim of this lint is to discourage reliance on this mechanism in normal code usage: it should only be used for backwards-compatibilty and the lint can be silenced by having users change their code. We can always later change this lint to be allowed by default if we consider that there are valid use cases for this feature other than backwards-compatiblity.
+
+### Type inference
+
+This change happens during name resolution and specifically doesn't interact with type inference. Consider this example:
```

Comment:
Name resolution is not where method selection is performed. Method resolution is performed during typechecking. 

I think the conclusion is still correct, though, but it has more to do with the order dependence of method selection and type inference; notice that this code works today:

```
trait Foo { fn method(&self) {} }
trait Bar: Foo { fn method(&self) {} }
impl<T> Foo for Vec<T> { }
impl<T: Copy> Bar for Vec<T> { }

fn main() {
    let mut x: Vec<Box<i32>> = vec![];
    // We disqualify `Bar`'s method since we know at this point
    // that `Vec<Box<i32>>` doesn't implement `Bar`, since
    // `Box<T>: !Copy`. Thus, there is no ambiguity.
    x.method();
    x.push(Box::new(22));
}
```